svcb.go

```diff
@@ -0,0 +1,661 @@
+				key_value = makeSvcKeyValue(SvcStringToKey(key))
+				if key_value == nil {
+					return &ParseError{"", "svc invalid key", l}
+				}
+			} else {
+				if idx == 0 {
+					return &ParseError{"", "no valid svc key found", l}
+				}
+				val = z[idx+1:]
+				key = z[0:idx]
+
+				key_value = makeSvcKeyValue(SvcStringToKey(key))
+				if key_value == nil {
+					return &ParseError{"", "svc invalid key", l}
+				}
+
+				if len(val) == 0 {
+					// We have a key and an equality sign
+					// Maybe we have nothing after "="
+					// or we have a double quote
+					l, _ = c.Next()
+					if l.value == zQuote {
+						l, _ = c.Next()
+						switch l.value {
+						case zString:
+							// We have a value in double quotes
+							val = l.token
+							l, _ = c.Next()
+							if l.value != zQuote {
+								return &ParseError{"", "svc unterminated value", l}
+							}
+						case zQuote:
+							// There's nothing in double quotes
+						default:
+							return &ParseError{"", "svc invalid value", l}
+						}
+					}
+				}
+			}
+			if err := key_value.Read(val); err != nil {
+				return &ParseError{"", err.Error(), l}
+			}
+			xs = append(xs, key_value)
+		case zQuote:
+			return &ParseError{"", "svc key can't contain double quotes", l}
+		case zBlank:
+		default:
+			return &ParseError{"", "bad svc Values", l}
+		}
+		l, _ = c.Next()
+	}
+	rr.Value = xs
+	if rr.Priority == 0 && len(xs) > 0 {
+		return &ParseError{"", "svc aliasform can't have values", l}
+	}
+	return nil
+}
+
+// makeSvcKeyValue returns a SvcKeyValue with the key
+// or nil if a reserved key is used.
+func makeSvcKeyValue(code uint16) SvcKeyValue {
+	switch code {
+	case SVCALPN:
+		return new(SVC_ALPN)
+	case SVCNO_DEFAULT_ALPN:
+		return new(SVC_NO_DEFAULT_ALPN)
+	case SVCPORT:
+		return new(SVC_PORT)
+	case SVCIPV4HINT:
+		return new(SVC_IPV4HINT)
+	case SVCESNICONFIG:
+		return new(SVC_ESNICONFIG)
+	case SVCIPV6HINT:
+		return new(SVC_IPV6HINT)
+	default:
+		if code == 0 || code == 65535 {
+			return nil
+		}
+		e := new(SVC_LOCAL)
+		e.Code = code
+		return e
+	}
+}
+
+// SVCB RR. TODO See RFC xxxx (https://tools.ietf.org/html/draft-ietf-dnsop-svcb-httpssvc-02)
+// The one with smallest priority SHOULD be given preference.
+// Of those with equal priority, a random one SHOULD be preferred for load balancing.
+type SVCB struct {
+	Hdr      RR_Header
+	Priority uint16
+	Target   string        `dns:"domain-name"`
+	Value    []SvcKeyValue `dns:"svc"` // if priority == 0 this is empty
+}
+
+// HTTPSSVCB RR. See the beginning of this file
+type HTTPSSVC struct {
+	SVCB
+}
+
+func (rr *HTTPSSVC) String() string {
+	return rr.SVCB.String()
+}
+
+func (rr *HTTPSSVC) parse(c *zlexer, o string) *ParseError {
+	return rr.SVCB.parse(c, o)
+}
+
+func (r1 *HTTPSSVC) isDuplicate(r2 RR) bool {
+	return (*r1).SVCB.isDuplicate(r2)
+}
+
+// SvcKeyValue defines a key=value pair for SvcFieldValue.
+// A SVCB RR can have multiple SvcKeyValues appended to it.
+type SvcKeyValue interface {
```

Comment:
oh wow. I'm reading section 6 of the draft, that's pretty terrible range of things you can put in this record as key=value thingies.
Note that the EDNS0 stuff has been shown to be suboptimal and isn't the nicest interface to work with, so I rather not have it here again. Having different types implementing SvcKeyValue works for me, so this will be still an interface. However things like `Key` and `Read` look like overkill. I.e. the interface must be paired down to the absolute minimal